src/normalize/exercise.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 OpenStax Poland
+// Licensed under the MIT license. See LICENSE file in the project root for
+// full license text.
+
+import { Editor, NodeEntry, Transforms } from 'slate'
+
+import { Commentary, Exercise, Problem, Section, Solution } from '../interfaces'
+import { normalizeOrderedChildren } from './util'
+
+/**
+ * Normalize an exercise or contents of an exercise
+ *
+ * Return true if entry was normalized and false otherwise.
+ */
+export default function normalizeExercise(editor: Editor, entry: NodeEntry): boolean {
+    const [node, path] = entry
+
+    // Problems, solutions, and commentaries make no sense outside an exercise.
+    // Should it happen however, just replace it with its contents.
+    if (Problem.isProblem(node) || Solution.isSolution(node)
+    || Commentary.isCommentary(node)) {
+        const [parent] = Editor.parent(editor, path)
+        if (!Exercise.isExercise(parent)) {
+            Transforms.unwrapNodes(editor, { at: path })
+            return true
+        }
+    }
+
+    if (Exercise.isExercise(node)) {
```

Comment:
I think `if (!Exercise.isExercise(node)) return false` is easier to read than long if statement where you have to look at indentation which is hard to see on Github review interface